angr/knowledge_plugins/sync/sync_controller.py

```diff
@@ -0,0 +1,376 @@
+                func_addr = arg.addr
+
+            # push many [comments, stack_vars]
+            elif isinstance(arg, list):
+                if isinstance(arg[0], SimStackVariable):
+                    func_addr = push_args[1].func_addr
+                elif isinstance(arg[0], binsync.data.Comment):
+                    func_addr = arg[0].func_addr
+
+            # push [comment]
+            elif isinstance(arg, int):
+                func_addr = arg #self.get_func_addr_from_addr(arg)
+
+            return func_addr
+
+        attr_func_addr = parse_push_args_func_addr(args)
+        attr_func_addr = attr_func_addr if attr_func_addr else -1
+
+        last_push_time = int(time.time())
+        last_push_func = attr_func_addr
+        #func_name = self._kb.functions[attr_func_addr].name if attr_func_addr in self._kb.functions else ""
+
+        f(self, *args, **kwargs)
+        self.client.last_push(last_push_func, last_push_time)
+
+    return set_last_push
+
+
+def make_state(f):
+    """
+    Build a writeable State instance and pass to `f` as the `state` kwarg if the `state` kwarg is None.
+    Function `f` should have have at least two kwargs, `user` and `state`.
+    """
+
+    @wraps(f)
+    def state_check(self, *args, **kwargs):
+        state = kwargs.pop('state', None)
+        user = kwargs.pop('user', None)
+        if state is None:
+            state = self.client.get_state(user=user)
+            kwargs['state'] = state
+            r = f(self, *args, **kwargs)
+            state.save()
+            return r
+        else:
+            kwargs['state'] = state
+            r = f(self, *args, **kwargs)
+            return r
+
+    return state_check
+
+
+def make_ro_state(f):
+    """
+    Build a read-only State instance and pass to `f` as the `state` kwarg if the `state` kwarg is None.
+    Function `f` should have have at least two kwargs, `user` and `state`.
+    """
+
+    @wraps(f)
+    def state_check(self, *args, **kwargs):
+        state = kwargs.pop('state', None)
+        user = kwargs.pop('user', None)
+        if state is None:
+            state = self.client.get_state(user=user)
+        kwargs['state'] = state
+        kwargs['user'] = user
+        return f(self, *args, **kwargs)
+
+    return state_check
+
+
+def init_checker(f):
+    @wraps(f)
+    def initcheck(self, *args, **kwargs):
+        if self.client is None:
+            raise ValueError("Please initialize SyncController by calling initialize(client).")
+        return f(self, *args, **kwargs)
+    return initcheck
+
+
+class SyncController(KnowledgeBasePlugin):
+    """
+    SyncController interfaces with a binsync client to push changes upwards and pull changes downwards.
+
+    :ivar binsync.Client client:   The binsync client.
+    """
+    def __init__(self, kb):
+        super().__init__()
+
+        self._kb: KnowledgeBasePlugin = kb
+        self.client: Optional[binsync.client.Client] = None
```

Comment:
This will fail if you don't have binsync imported. Use `'binsync.client.Clinet'` instead.